src/Controllers/UserController.js

```diff
@@ -45,14 +49,56 @@ export class UserController extends AdaptableController {
       // TODO: Better error here.
       throw undefined;
     }
-    return this.config.database.update('_User', {
+
+    let query = {username: username, _email_verify_token: token};
+
+    let updateFields = { emailVerified: true, _email_verify_token: {__op: 'Delete'}};
+    if (this.config.allowEmailVerifyTokenToExpire) {
+      updateFields._email_verify_token_expires_at = {__op: 'Delete'};
+    }
+
+    return this.checkEmailTokenValidity(username, token).then((user) => {
+      return this.config.database.update('_User', query, updateFields).then((document) => {
+        if (!document) {
+          throw undefined;
+        }
+        return Promise.resolve(document);
+      });
+
+    });
+  }
+
+  checkEmailTokenValidity(username, token) {
+    // don't check for token validity
+    if (!this.config.allowEmailVerifyTokenToExpire) {
+      return Promise.resolve();
+    }
+
+    return this.config.database.find('_User', {
       username: username,
       _email_verify_token: token
-    }, {emailVerified: true}).then(document => {
-      if (!document) {
+    }, {limit: 1}).then(results => {
```

Comment:
Sounds good. I will incorporate your suggestion. I will add `_email_verify_expires_at` checking in the query for the `update` statement.

btw: You may know this already...if there is a failure then we redirect the user to the `invalid_link`. We currently don't send the error message/failure reason back with the `invalid_link`. If there is a request to send the failure reason then I will take a shot at it in another PR.